ethcore/src/private_transactions/mod.rs

```diff
@@ -0,0 +1,907 @@
+		tx.gas_required(&self.client.latest_schedule()).into()
+	}
+
+	fn is_service_transaction_acceptable(&self, _tx: &SignedTransaction) -> Result<bool, String> {
+		Ok(false)
+	}
+}
+
+/// Configurtion for private transaction provider
+#[derive(Default, PartialEq, Debug)]
+pub struct ProviderConfig {
+	/// Accounts that can be used for validation
+	pub validator_accounts: Vec<Address>,
+	/// Account used for signing public transactions created from private transactions
+	pub signer_account: Option<Address>,
+	/// Passwords used to unlock accounts
+	pub passwords: Vec<String>,
+	/// Account used for signing requests to key server
+	pub key_server_account: Option<Address>,
+	/// URL to key server
+	pub key_server_url: Option<String>,
+	/// Key server's threshold
+	pub key_server_threshold: u32,
+}
+
+#[derive(Debug)]
+/// Private transaction execution receipt.
+pub struct Receipt {
+	/// Private transaction hash.
+	pub hash: H256,
+	/// Created contract address if any.
+	pub contract_address: Option<Address>,
+	/// Execution status.
+	pub status_code: u8,
+}
+
+/// Manager of private transactions
+pub struct Provider {
+	encryptor: RwLock<Arc<Encryptor>>,
+	config: RwLock<ProviderConfig>,
+	notify: RwLock<Vec<Weak<ChainNotify>>>,
+	transactions_for_signing: Mutex<SigningStore>,
+	transactions_for_verification: Mutex<VerificationStore>,
+	client: RwLock<Option<Weak<Client>>>,
+	accounts: RwLock<Option<Weak<AccountProvider>>>,
+}
+
+#[derive(Debug)]
+struct PrivateExecutionResult<T, V> where T: Tracer, V: VMTracer {
+	code: Option<Bytes>,
+	state: Bytes,
+	result: Executed<T::Output, V::Output>,
+}
+
+impl Provider where {
+	/// Create a new provider.
+	pub fn new() -> Result<Self, EthcoreError> {
+		Ok(Provider {
+			config: RwLock::new(ProviderConfig::default()),
+			notify: RwLock::new(Vec::new()),
+			transactions_for_signing: Mutex::new(SigningStore::new()),
+			transactions_for_verification: Mutex::new(VerificationStore::new()),
+			client: RwLock::new(None),
+			accounts: RwLock::new(None),
+			encryptor: RwLock::new(Arc::new(SecretStoreEncryptor::empty()?)),
+		})
+	}
+
+	/// Adds an actor to be notified on certain events
+	pub fn add_notify(&self, target: Arc<ChainNotify>) {
+		self.notify.write().push(Arc::downgrade(&target));
+	}
+
+	fn notify<F>(&self, f: F) where F: Fn(&ChainNotify) {
+		for np in self.notify.read().iter() {
+			if let Some(n) = np.upgrade() {
+				f(&*n);
+			}
+		}
+	}
+
+	/// Register client reference
+	pub fn register_client(&self, client: Weak<Client>) {
+		*self.client.write() = Some(client);
+	}
+
+	/// Register accounts provider
+	pub fn register_account_provider(&self, accounts: Weak<AccountProvider>) {
+		*self.accounts.write() = Some(accounts);
+	}
+
+	/// Sets encryptor
+	pub fn set_encryptor(&self, encryptor: Arc<Encryptor>) {
+		*self.encryptor.write() = encryptor.clone();
+	}
+
+	/// Sets provider's config.
+	pub fn set_config(&self, config: ProviderConfig) -> Result<(), EthcoreError> {
```

Comment:
Is there a need to be able to `set_config` after creation? Why not just pass config to `new`?